Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,313 @@
+
+type_argument_u
+    : type | '_'
+    ;
+
+```
+
+To be able to use it in invocation, we have to modify the method invocation and type inference algorithm.
+
+> #### 12.8.9.2 Method invocations
+>
+> ...
+>
+> - **If F is generic and M includes a type argument list containing at least one `_` placeholder, F is a candidate when:**
+>   - **F has the same number of method type parameters as were supplied in the type argument list, and**
+>   - **Type inference (§12.6.3) succeeds, inferring a list of type arguments for the call, and**
+>   - **Once the type arguments are substituted for the corresponding method type parameters, all constructed types in the parameter list of F satisfy their constraints (§8.4.5), and the parameter list of F is applicable with respect to A (§12.6.4.2).**
+> - If F is generic and M includes a type argument list, F is a candidate when:
+
+> #### 12.6.3.1 General
+> 
+> ...
+> 
+> When a particular method group is specified in a method invocation, and either no type arguments are specified as part of the method invocation or the type argument list contains at least one '_', type inference is applied to each generic method in the method group.
+>
+> ...
+>
+> With a method call of the form `M(E₁ ...Eₓ)` **or `M<Y₁...Yᵥ>(E₁ ... Eₓ)`** the task of type inference is to find unique type arguments `S₁...Sᵥ` for each of the type parameters `X₁...Xᵥ` so that the call `M<S₁...Sᵥ>(E₁...Eₓ)` becomes valid **and in the case of `M<Y₁...Yᵥ>(E₁ ... Eₓ)` type argument `Sk` should by identical with `Yk` where Yk is not `_`**.
+>
+> ...
+>
+> #### 12.6.3.2 The first phase
+> **If the method call has the form `M<Y₁...Yᵥ>(E₁ ... Eₓ)` for each `Yᵢ` which is not `_`, fix type variable `Xᵢ` to `Yᵢ`.**
+> For each of the method arguments `Eᵢ`:
+
+If we consider dynamic overload resolution, we modify compile-time checks to reflect the change.
+
+> 12.6.5 Compile-time checking of dynamic member invocation
+>
+> First, if F is a generic method and type arguments were provided, then those **which are not `_`** are substituted for the type parameters in the parameter list. However, if type arguments were not provided, no such substitution happens.
+
+
+There raises the question if we can go further and allow `_` to be nested 
+See the following example.
+
+```csharp
+TResult GetResult<TResult>() where TResult : IEnumerable<string> {}
+
+var results = GetResult<List<_>>(); // Just specifying wrapper implementation.
+```
+
+It would be most beneficial during working with any generic wrappers, where we are curious about the wrapper implementation, but not which will be inside the wrapper.
+
+This kind of wildcard needs more information about the relation between type arguments to become useful.
+More information can be extracted from generic type constraints, which can bring also more inference power to simple `_` inferred types.
+Unfortunately, using information about constraints in type inference causes breaking change, see the following example.
```

Comment:
I think we should treat *constraints contributing to type inference* as a separate feature. We should consider doing it.